pulp_python/app/models.py

```diff
@@ -83,18 +97,171 @@ def publish(self):
+        """
+
+        package = {}
+
+        package['filename'] = distribution['filename']
+        package['packagetype'] = distribution['packagetype']
+        package['name'] = project['name']
+        package['version'] = version
+        package['metadata_version'] = project.get('metadata_version')
+        package['summary'] = project.get('summary')
+        package['description'] = project.get('description')
+        package['keywords'] = project.get('keywords')
+        package['home_page'] = project.get('home_page')
+        package['download_url'] = project.get('download_url')
+        package['author'] = project.get('author')
+        package['author_email'] = project.get('author_email')
+        package['maintainer'] = project.get('maintainer')
+        package['maintainer_email'] = project.get('maintainer_email')
+        package['license'] = project.get('license')
+        package['requires_python'] = project.get('requires_python')
+        package['project_url'] = project.get('project_url')
+        package['platform'] = project.get('platform')
+        package['supported_platform'] = project.get('supported_platform')
+        package['requires_dist'] = project.get('requires_dist', [])
+        package['provides_dist'] = project.get('provides_dist', [])
+        package['obsoletes_dist'] = project.get('obsoletes_dist', [])
+        package['requires_external'] = project.get('requires_external', [])
+        package['url'] = distribution['url']
+        package['md5_digest'] = distribution['md5_digest']
+
+        return package
+
+    def _build_additions(self, additions, remote_metadata):
+        """
+        Generate the content to be added.
+
+        Returns:
+            generator: A generator of content to be added.
+        """
+        for entry in remote_metadata:
+            if(entry['filename'] not in additions):
+                continue
+
+            url = entry.pop('url')
+            artifact = Artifact(md5=entry.pop('md5_digest'))
+
+            package = PythonPackageContent(**entry)
+            content = PendingContent(
+                package,
+                artifacts={
+                    PendingArtifact(model=artifact, url=url, relative_path=entry['filename'])
+                })
+            yield content
+
+    def _build_removals(self, removals):
+        """
+        Generate the content to be removed.
+
+        Returns:
+            generator: A generator of content to be removed.
+        """
+        q = models.Q()
+        for content_key in removals:
+            q |= models.Q(pythonpackagecontent__filename=content_key)
+            q_set = self.repository.content.filter(q)
+            q_set = q_set.only('id')
+            for content in q_set:
+                yield content
 
     def sync(self):
         """
-        Synchronize the repository with the remote repository.
+        Sync the python projects listed on a :class:`pulp_python.app.models.Serializer`
```

Comment:
I don't think you need to mention the serializer, the project list stored is on this object.

Reply:
oops I totally meant importer there